### PR TITLE
Fish With Version

### PR DESCRIPTION
**Pull request by jafeltra**
This PR supports fishing for definitions with a version. This PR adds support for fishing with a specified version in the `FSHTank` and in the `Package`. Support for fishing with a specified version in `FHIRDefinitions` is provided by FPL. There is an FPL PR with this support [here](https://github.com/standardhealth/fhir-package-loader/pull/11).

In addition to supporting fishing with a version, there are a number of places that SUSHI tries to fish for the version specified but is satisfied if any version is found, but logs a warning. To support this, three util functions are included: `fishForFHIRBestVersion`, `fishForMetadataBestVersion`, and `fishInTankBestVersion`. These functions will call either `fishForFHIR`, `fishForMetadata`, or `fish` (in a Tank) first with the item and specified version. If nothing is found, they'll then try again without a version, and if a definition/metadata is found with a differing version, a warning is logged. I replaced the special case handling SUSHI was doing to slice off the version and check for discrepancies between versions with these functions where possible.

To test this, you can use the FPL PR (until a release is ready and available directly in this PR) and link your local FPL to SUSHI. Then you can run the tests and test FSH projects locally.

I ran a full regression and found that two repos had regressions, [HL7/vrdr](https://github.com/HL7/vrdr) and [nightingaleproject/vital_records_fhir_messaging_ig](https://github.com/nightingaleproject/vital_records_fhir_messaging_ig), both with the same issue. Both repos specify a line similar to:

```
* valueDateTime.extension[partialDateTime].extension[time].valueTime.extension[dataabsent].valueCode = http://hl7.org/fhir/ValueSet/data-absent-reason|4.0.1#unknown
```
This line is setting a URL with a version as a system in a FSH code. Before, SUSHI was just fishing for this, didn't find a definition so didn't realize it was a ValueSet, checked that it was a valid URI, and successfully used the code system. In this branch, fishing with the version will work correctly and SUSHI will find a ValueSet definition for this definition, which is a problem, and it will throw a MismatchBindingTypeError and not set the value. I think this regression is actually the correct behavior for SUSHI now.

~_Note: this PR is marked as draft until a new version of FPL is released with support for fishing with versions so that the FHIR Definitions fishing works correctly._~ FPL v.0.3.0 is used now, so this is no longer draft!

### CODE REVIEW

File: src/export/StructureDefinitionExporter.ts
Comment (cmoesel) [inline src/export/StructureDefinitionExporter.ts:815]: I think this is another case where a variant of my regex replace trick might work.
```
const profileUrl = item.type.replace(/^[^|]+/, extension.url);
```

File: src/export/StructureDefinitionExporter.ts
Comment (cmoesel) [reply src/export/StructureDefinitionExporter.ts:622]: Also -- this is one of the places we didn't preserve the version before and now we do. I think that's a good idea. Thanks!

File: src/export/ValueSetExporter.ts
Comment (cmoesel) [inline src/export/ValueSetExporter.ts:60]: I see that this change makes it more consistent w/ the approach we take for value sets below -- so that's good.

I also see that if there is a version, we drop the version when there's a match but keep it when there isn't (value sets below does the same thing). I think this is one of those places where we want to keep the version when there is a match. I think we had decided, however, that these kinds of changes aren't necessarily in scope for this PR and can be deferred to another PR if we want.

File: src/fhirtypes/common.ts
Comment (cmoesel) [inline src/fhirtypes/common.ts:867]: Another opportunity for the regex replace (I think):
```ts
assignedCode.system = value.system.replace(/^[^|]+/, codeSystemMeta.url);
```

File: src/fhirtypes/common.ts
Comment (cmoesel) [inline src/fhirtypes/common.ts:1323]: An `Invariant` will never have a version in it. Nor will a `Mapping`.  I guess a `RuleSet` might, but I'm not sure that's relevant here since you don't know where the RuleSet is applied?

File: src/import/FSHTank.ts
Comment (cmoesel) [inline src/import/FSHTank.ts:366]: Invariants don't carry versions, so I don't think this is needed.

File: src/utils/FishingUtils.ts
Comment (cmoesel) [inline src/utils/FishingUtils.ts:28]:
```suggestion
  if (result == null && item?.includes('|')) {
```

File: src/utils/FishingUtils.ts
Comment (cmoesel) [inline src/utils/FishingUtils.ts:32]:
```suggestion
    if (version != null && result?.version != null && version != result.version) {
```

File: src/utils/FishingUtils.ts
Comment (cmoesel) [inline src/utils/FishingUtils.ts:56]:
```suggestion
  if (metadata == null && item?.includes('|')) {
```

File: src/utils/FishingUtils.ts
Comment (cmoesel) [inline src/utils/FishingUtils.ts:60]:
```suggestion
    if (version != null && metadata?.version != null && version != metadata.version) {
```

File: src/utils/FishingUtils.ts
Comment (cmoesel) [inline src/utils/FishingUtils.ts:34]: I feel like the language is not quite right here because I initially interpreted "The ${base} definition was specified with version ${version}" to mean it was _defined_ with version ${version}. But what we really mean is it was _requested_ with that version. So maybe:

> The ${base} definition was requested with version ${version}, but SUSHI found version ${result.version}

or even just putting the original value with the `|` in it, e.g.:

> ${item} was requested, but SUSHI found ${base}|${result.version}

File: src/utils/FishingUtils.ts
Comment (cmoesel) [inline src/utils/FishingUtils.ts:62]: Whatever language you decide on for above should be reflected here too.

File: src/utils/FishingUtils.ts
Comment (cmoesel) [inline src/utils/FishingUtils.ts:91]:
```suggestion
  if (result == null && item?.includes('|')) {
```

File: src/utils/FishingUtils.ts
Comment (cmoesel) [inline src/utils/FishingUtils.ts:99]: And again, make sure the final language gets reflected here.

File: test/export/InstanceExporter.test.ts
Comment (cmoesel) [inline test/export/InstanceExporter.test.ts:5900]: I don't think a `CodeSystem` definition would declare its `url` with a version. The `url` would be the version-less URL and the `version` element would contain the version. E.g.
```ts
const visibleSystemUrl = new CaretValueRule('');
visibleSystemUrl.caretPath = 'url';
visibleSystemUrl.value = 'http://hl7.org/fhir/us/minimal/CodeSystem/Visible';
const visibleSystemVersion = new CaretValueRule('');
visibleSystemVersion.caretPath = 'version';
visibleSystemVersion.value = '1.0.0';
```

File: test/import/FSHTank.test.ts
Comment (cmoesel) [inline test/import/FSHTank.test.ts:236]: I think you can delete this since invariants don't have a version.

File: test/import/FSHTank.test.ts
Comment (cmoesel) [inline test/import/FSHTank.test.ts:787]: I think these can be removed. Also, Invariants don't support caret rules anyway. :-)

**Review by cmoesel — CHANGES_REQUESTED** [review]
Wow. This turned out to be pretty intense! But... we finally have a more consistent approach to handling versioned canonicals! Thank you!

I found a few things that I think can be improved and/or fixed and/or removed. But... this is looking very good!

File: src/export/Package.ts
Comment (jafeltra) [reply src/export/Package.ts:54]: Well... I thought there was a reason, but now I can't think of what that reason was.

Comment (cmoesel) [reply src/export/Package.ts:54]: I think `(version == null || (i.version && i.version === version))` can be shortened to `(version == null || i.version === version)` because:
* If `version` is null or undefined, it is true without need to evaluate the 2nd part;
* If `version` is not null/undefined, then it won't be equal to `i.version` if `i.version` is null anyway.

File: src/export/ValueSetExporter.ts
Comment (jafeltra) [reply src/export/ValueSetExporter.ts:60]: Now that I have your nifty regex replace in my back pocket, I don't mind doing it here. But I can also hold off for a separate PR if that's preferred.

Comment (jafeltra) [reply src/export/ValueSetExporter.ts:60]: Actually, wait, now I'm not sure if I agree we should keep the version. The version the author specified gets added to the `composeElement.version` regardless of if we find a match on line 63. And because the `foundSystem` variable does a final `split('|')` and then only `foundSystem[0]` is used later, the version is always spliced off and never added to `composeElement.system`. So do we want to include the `|version` on the `system` property, as well as on the `version` property?

File: src/export/Package.ts
Comment (jafeltra) [reply src/export/Package.ts:54]: Yeah agreed. Maybe I had something different at some point and changed things around and didn't realize the redundancy in the end.

File: src/fhirtypes/common.ts
Comment (jafeltra) [reply src/fhirtypes/common.ts:867]: I wonder if it would be convenient or annoying to wrap the regex replace in a helper function...

File: test/export/InstanceExporter.test.ts
Comment (jafeltra) [reply test/export/InstanceExporter.test.ts:5900]: I thought I had added this case because of a regression, but I'm not sure which repo without rerunning the regression. I believe this was also tied to [this](https://github.com/FHIR/sushi/pull/1275#discussion_r1196899479) change, and updating to use the regex replace actually causes this test to fail. I guess it is worth keeping the test with the version caret value rule, so I'll update this!

File: src/utils/FishingUtils.ts
Comment (jafeltra) [reply src/utils/FishingUtils.ts:34]: I like the simple second version!

File: src/export/Package.ts
Comment (jafeltra) [reply src/export/Package.ts]: Addressed in [9668cdc](https://github.com/FHIR/sushi/pull/1275/commits/9668cdc2e8174378a573a33170c4740424a1b2cd)

File: src/export/StructureDefinitionExporter.ts
Comment (jafeltra) [reply src/export/StructureDefinitionExporter.ts]: Part of [70e31fc](https://github.com/FHIR/sushi/pull/1275/commits/70e31fc1b4bdb5eaab6119f21d08e44e038c24db)

File: src/fhirtypes/common.ts
Comment (jafeltra) [reply src/fhirtypes/common.ts]: Part of [70e31fc](https://github.com/FHIR/sushi/pull/1275/commits/70e31fc1b4bdb5eaab6119f21d08e44e038c24db).

And I'm leaning that it would be annoying to wrap in a helper function... so didn't do that.

File: src/import/FSHTank.ts
Comment (jafeltra) [reply src/import/FSHTank.ts]: Part of [0d0852d](https://github.com/FHIR/sushi/pull/1275/commits/0d0852d19cbf5a24e3e63fdf9a7f359ac288bcce)

File: src/utils/FishingUtils.ts
Comment (jafeltra) [reply src/utils/FishingUtils.ts]: Part of [18b8ba8](https://github.com/FHIR/sushi/pull/1275/commits/18b8ba89e0c6aa3a9ec12bfc0c288154ec6be7f5)

File: src/utils/FishingUtils.ts
Comment (jafeltra) [reply src/utils/FishingUtils.ts]: Part of [18b8ba8](https://github.com/FHIR/sushi/pull/1275/commits/18b8ba89e0c6aa3a9ec12bfc0c288154ec6be7f5)

File: src/utils/FishingUtils.ts
Comment (jafeltra) [reply src/utils/FishingUtils.ts]: Part of [18b8ba8](https://github.com/FHIR/sushi/pull/1275/commits/18b8ba89e0c6aa3a9ec12bfc0c288154ec6be7f5)

File: src/utils/FishingUtils.ts
Comment (jafeltra) [reply src/utils/FishingUtils.ts]: Part of [18b8ba8](https://github.com/FHIR/sushi/pull/1275/commits/18b8ba89e0c6aa3a9ec12bfc0c288154ec6be7f5)

File: src/utils/FishingUtils.ts
Comment (jafeltra) [reply src/utils/FishingUtils.ts]: Simple second version updated in Part of [18b8ba8](https://github.com/FHIR/sushi/pull/1275/commits/18b8ba89e0c6aa3a9ec12bfc0c288154ec6be7f5). Let me know if you prefer something else!

File: src/utils/FishingUtils.ts
Comment (jafeltra) [reply src/utils/FishingUtils.ts]: Part of [18b8ba8](https://github.com/FHIR/sushi/pull/1275/commits/18b8ba89e0c6aa3a9ec12bfc0c288154ec6be7f5)

File: src/utils/FishingUtils.ts
Comment (jafeltra) [reply src/utils/FishingUtils.ts:87]: Part of [18b8ba8](https://github.com/FHIR/sushi/pull/1275/commits/18b8ba89e0c6aa3a9ec12bfc0c288154ec6be7f5)

File: src/utils/FishingUtils.ts
Comment (jafeltra) [reply src/utils/FishingUtils.ts]: Part of [18b8ba8](https://github.com/FHIR/sushi/pull/1275/commits/18b8ba89e0c6aa3a9ec12bfc0c288154ec6be7f5) (although you must know that by now ;) )

File: test/export/InstanceExporter.test.ts
Comment (jafeltra) [reply test/export/InstanceExporter.test.ts]: Included in [70e31fc](https://github.com/FHIR/sushi/pull/1275/commits/70e31fc1b4bdb5eaab6119f21d08e44e038c24db)

File: test/import/FSHTank.test.ts
Comment (jafeltra) [reply test/import/FSHTank.test.ts]: Part of [0d0852d](https://github.com/FHIR/sushi/pull/1275/commits/0d0852d19cbf5a24e3e63fdf9a7f359ac288bcce)

File: test/import/FSHTank.test.ts
Comment (jafeltra) [reply test/import/FSHTank.test.ts]: Part of [0d0852d](https://github.com/FHIR/sushi/pull/1275/commits/0d0852d19cbf5a24e3e63fdf9a7f359ac288bcce)

File: src/export/ValueSetExporter.ts
Comment (cmoesel) [reply src/export/ValueSetExporter.ts]: Oh, dang. You're right! For whatever reason, `ValueSet.compose.include` splits the version off into its own element. And the `system` element is `uri` _not_ `canonical`. So, yep, we definitely do _not_ want to attach the `|version`.  Good call!

File: test/export/InstanceExporter.test.ts
Comment (cmoesel) [reply test/export/InstanceExporter.test.ts]: Technically, `CodeSystem.url` is a `uri`, so you're not supposed to append a version to it. It even says:
> URL should not contain | or # - these characters make processing canonical references problematic

So if someone is making a CodeSystem with a URL like that in a real IG, we shouldn't support it anyway.

Comment (jafeltra) [reply test/export/InstanceExporter.test.ts]: Oh! Well maybe I just invented that test case. It seems so specific so I thought I must have pulled it from somewhere, but maybe not. Well, it is no longer supported now!

**Review by cmoesel — APPROVED** [review]
This looks very good! I ran a regression and got the same two regressions as you did. Upon studying them, however, I realized that they're assigning the value to a `code` -- so while the system in the FSH is incorrect, it isn't actually _used_ because primitive `code` does not include the system. This means the assignment can still safely happen.

That being the case, I thought it would be more appropriate to detect that situation and _warn_ on it rather than _error_ -- and proceed with the assignment rather than skipping it. So that is what I've done in 3a9f3e9.

Now everything looks good to me and has my approval!

File: src/fhirtypes/common.ts
Comment (mint-thompson) [inline src/fhirtypes/common.ts:859]: This is not really a problem, but it's possible for each of these functions to separately log a warning for a version mismatch if the CodeSystem is retrieved from the package when fishing for metadata. Here's an example that can make this happen:
```
CodeSystem: BreakfastCS
* ^version = "1.0.0"
* #toast "Toast"
* #donut "Donut"

Instance: MyBreakfast
InstanceOf: Observation
* status = #final
* code = BreakfastCS|2.0.0#toast
```

**Review by mint-thompson — APPROVED** [review]
This looks great, thank you! There's one slightly unusual thing that I noticed, but I don't think it necessarily needs to be changed, so I'm giving my approval.

File: src/fhirtypes/common.ts
Comment (cmoesel) [reply src/fhirtypes/common.ts:859]: Ah, yeah. We technically need to do both fishes -- so the only way I can think of to avoid the double warning is to do the 2nd one based on whatever was actually returned by the first (rather than the original request). E.g.,:

```ts
    const codeSystemMeta = fishForMetadataBestVersion(fisher, value.system, Type.CodeSystem);
    if (codeSystemMeta) {
      clone = cloneDeep(rule);
      const assignedCode = getRuleValue(clone) as FshCode;
      assignedCode.system = value.system.replace(/^[^|]+/, codeSystemMeta.url);

      // Find the code system using the returned metadata to avoid duplicate warnings if version mismatches
      const matchedCanonical = codeSystemMeta.url
        ? `${codeSystemMeta.url}${codeSystemMeta.version ? `|${codeSystemMeta.version}` : ''}`
        : value.system;
      const codeSystem = fishInTankBestVersion(tank, matchedCanonical, Type.CodeSystem);
      if (codeSystem && (codeSystem instanceof FshCodeSystem || codeSystem instanceof Instance)) {
        // if a local system was used, check to make sure the code is actually in that system
        listUndefinedLocalCodes(codeSystem, [assignedCode.code], tank, rule);
      }
    }
```

@mint-thompson - I have this change locally and it _does_ worth. Do you think I should push it up?

Comment (mint-thompson) [reply src/fhirtypes/common.ts:859]: I think so, yes. It's not too complicated, and it makes the output more consistent. Thank you!

Comment (cmoesel) [reply src/fhirtypes/common.ts]: OK. Fixed in 858c75d. I also noticed that the warning messages did not log the source info -- so we'd say there was a mismatch, but it would be up to the author to find it in their FSH files. This didn't seem optimal, so I also updated it to pass along the source info and log that. It's in the same commit.

**Review by cmoesel — COMMENTED** [review]
OK. I've addressed the double-warning and source info logging. @mint-thompson - can you give it one more review?

**Review by mint-thompson — APPROVED** [review]

**Pull request merged**
